### PR TITLE
Test openreg inference hooks

### LINKED ISSUE #172321: [OpenReg] Extend test_autograd.py coverage
### 🚀 The feature, motivation and pitch

Extend test_autograd.py (currently only 1 test) with comprehensive autograd testing to validate gradient computation and automatic differentiation on OpenReg devices.

  #### Background

  Autograd is a core component of PyTorch that enables automatic differentiation. For any backend to be useful for training workloads, it must correctly integrate with PyTorch's autograd engine.

  The current `test_autograd.py` has only 1 test, which is insufficient to validate:
  - Gradient computation correctness
  - Interaction with autograd context managers
  - Complex backward pass scenarios

  Extending this coverage ensures backend authors understand what autograd behaviors to implement and validate.

Parent issue: [169597](https://github.com/pytorch/pytorch/issues/169597)

  ### Work Items

  - [ ] Gradient accumulation across multiple backward calls
  - [ ] Backward with `retain_graph=True`
  - [ ] Operations under `torch.no_grad()`
  - [ ] Operations under `torch.inference_mode()`
  - [ ] Verify `requires_grad` propagates correctly across operations
  - [ ] Verify `tensor.detach()` behavior
  - [ ] Tensor and module backward hooks on device

  @parsshar-RH @ahkush @groenenboomj

### Alternatives

_No response_

### Additional context

  Related files:
  - `test/openreg/test_autograd.py`
  - PyTorch autograd documentation: https://pytorch.org/docs/stable/autograd.html

cc @seemethere @malfet @pytorch/pytorch-dev-infra

### PR DESCRIPTION
**Pull request by CloseChoice**
Fixes #172321

A lot of what was required in the issue was already solved in https://github.com/pytorch/pytorch/commit/b76165aa64c548ed4a69e9f837ef48752b93be9a. This PR just adds the two missing tests from #172321:
 - test_autograd_inference_mode
 - test_backward_hook

cc [mansiag05](https://github.com/mansiag05) @parsshar-RH

### CODE REVIEW

File: test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_autograd.py [hunk 1]
```diff
@@ -75,6 +75,19 @@ def test_autograd_no_grad(self):
...
         self.assertTrue(z.requires_grad)
 
+    def test_autograd_inference_mode(self):
+        """Test inference_mode context manager"""
+        x = torch.randn(2, 3, device="openreg", requires_grad=True)
+
+        with torch.inference_mode():
+            y = x * 2
+            self.assertIsNone(y.grad_fn)
+            self.assertFalse(y.requires_grad)
```
Comment (mansiag05) [inline test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_autograd.py:85]: test_autograd_inference_mode is analogous to test_autograd_no_grad. Since inference_mode has stronger semantics (inference tensors can't be used in autograd outside the context), would it be worth adding an assertion that verifies this?

File: test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_autograd.py [hunk 2]
```diff
@@ -121,6 +134,37 @@ def test_autograd_cross_device(self):
         self.assertIsNotNone(x.grad)
         self.assertTrue(x.grad.is_cpu)
 
+    def test_backward_hook(self):
```
Comment (mansiag05) [inline test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_autograd.py:144]: A couple of suggestions on test_backward_hook:

  1. nit: The forward correctness assertions (result == a * b, result.device.type) don't depend on backward(), could you move them before the backward() so that a forward-pass failure produces a clearer error message?
  2. The test verifies that hooks are called on the right device, but doesn't check gradient values. Would it be worth adding a check to validate correctness?

**Review by mansiag05 — CHANGES_REQUESTED** [review]

File: test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_autograd.py (hunk 2 above)
Comment (CloseChoice) [reply test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_autograd.py:144]: Added code to store the gradient and verify the value is correct

File: test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_autograd.py (hunk 1 above)
Comment (CloseChoice) [reply test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_autograd.py:85]: done.

**Review by CloseChoice — COMMENTED** [review]
Thanks for the review and sorry for the very late reply. Hope this is still helpful. Let me know if anything else comes up here.
